src/Extension/Engine/SiteTreePublishingEngine.php

```diff
@@ -36,130 +42,248 @@ class SiteTreePublishingEngine extends SiteTreeExtension implements Resettable
+            // until the publish action has completed successfully, and onAfterPublishRecursive() has been called. This
+            // is because we don't want to queue jobs if the publish action fails
+            $this->collectChanges($context);
         }
+    }
+
+    /**
+     * @param SiteTree|SiteTreePublishingEngine|null $original
+     */
+    public function onAfterPublishRecursive($original)
+    {
+        // Flush any/all changes that we might have collected from onBeforePublishRecursive()
+        $this->flushChanges();
+
+        $parentId = $original->ParentID ?? null;
+        $urlSegment = $original->URLSegment ?? null;
+
+        $owner = $this->getOwner();
+
+        // Apparently ParentID can sometimes be string, so make sure we cast to (int) for our comparison
+        $parentChanged = $parentId && (int) $parentId !== (int) $owner->ParentID;
+        $urlSegmentChanged = $urlSegment && $original->URLSegment !== $owner->URLSegment;
+
         $context = [
-            'action' => 'publish',
+            'action' => self::ACTION_PUBLISH,
+            // If a URL change has been detected, then we need to force the recursive regeneration of all child
+            // pages
+            'urlSegmentChanged' => $parentChanged || $urlSegmentChanged,
         ];
-        $this->collectChanges($context);
+
+        // Collect any additional changes (noting that some could already have been added in onBeforePublishRecursive())
+        // This collectChanges will use the stage of DRAFT. This is purely because if a page has an unpublished parent
+        // then the LIVE URL will be incorrect (the child page URL would be in the root slug) - we'd prefer to cache to
+        // correct URL (with parentage intact) even though it'll be a cache of a 404
+        $this->collectChanges($context, Versioned::DRAFT);
+        // Flush any/all changes that we have detected
         $this->flushChanges();
     }
 
     public function onBeforeUnpublish()
     {
         $context = [
-            'action' => 'unpublish',
+            'action' => self::ACTION_UNPUBLISH,
         ];
+        // We'll collect these changes now, but they won't be actioned until onAfterUnpublish()
+        // This collectChanges will use the stage of LIVE, because we specifically need to retrieve the current LIVE
+        // URLs for the page before the unpublish completes
         $this->collectChanges($context);
     }
 
     public function onAfterUnpublish()
     {
+        // Flush any/all changes that we have detected
         $this->flushChanges();
     }
 
     /**
      * Collect all changes for the given context.
      *
      * @param array $context
+     * @param string $stage The Stage (LIVE/DRAFT) that the URLs will be requested from
+     * @return void
      */
-    public function collectChanges($context)
+    public function collectChanges($context, string $stage = Versioned::LIVE)
```

Comment:
We still can't add new default arguments to public or protected methods, for backwards compatability reasons. :p

Reply:
Ok... I can do it another way, the logic is just getting really convoluted and difficult to track.

Reply:
@GuySartorelli @emteknetnz can I please check in on this one again. I kinda thought you could safely add new default arguments to methods and not break BC - unless the method is an `interface` (which this isn't).

I could be wrong, in which case, I'll leave the code as it is now, but I think this version was more readable and easy to understand than the workaround I had to add.

Reply:
You can't add them unfortunately, because any class which _subclasses_ this one would have to update its method signature. We've learned that the hard way.